sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala

```diff
@@ -352,3 +380,142 @@ case class BroadcastPartitioning(mode: BroadcastMode) extends Partitioning {
     case _ => false
   }
 }
+
+/**
+ * This is used in the scenario where an operator has multiple children (e.g., join) and one or more
+ * of which have their own requirement regarding whether its data can be considered as
+ * co-partitioned from others. This offers APIs for:
+ *
+ *   1. Comparing with specs from other children of the operator and check if they are compatible.
+ *      When two specs are compatible, we can say their data are co-partitioned, and Spark will
+ *      potentially able to eliminate shuffle if necessary.
+ *   1. Creating a partitioning that can be used to re-partition another child, so that to make it
+ *      having a compatible partitioning as this node.
+ */
+trait ShuffleSpec {
+  /**
+   * Returns the number of partitions of this shuffle spec
+   */
+  def numPartitions: Int
+
+  /**
+   * Returns true iff this spec is compatible with the other [[Partitioning]] and
+   * clustering expressions (e.g., from [[ClusteredDistribution]]).
+   *
+   * A true return value means that the data partitioning from this spec can be seen as
+   * co-partitioned with the `otherPartitioning`, and therefore no shuffle is required when
+   * joining the two sides.
+   */
+  def isCompatibleWith(other: ShuffleSpec): Boolean
+
+  /**
+   * Creates a partitioning that can be used to re-partitioned the other side with the given
+   * required distribution.
+   *
+   * Note: this will only be called after `isCompatibleWith` returns true on the side where the
+   * `clustering` is returned from.
+   */
+  final def createPartitioning(distribution: Distribution): Partitioning = distribution match {
+    case AllTuples =>
+      SinglePartition
+    case ClusteredDistribution(clustering, _) =>
+      createPartitioning0(clustering)
+    case _ =>
+      throw new IllegalStateException("unexpected distribution: " +
+          s"${distribution.getClass.getSimpleName}")
+  }
+
+  def createPartitioning0(clustering: Seq[Expression]): Partitioning
+}
+
+case object SinglePartitionShuffleSpec extends ShuffleSpec {
+  override def isCompatibleWith(other: ShuffleSpec): Boolean = {
```

Comment:
My understanding is, if two partitionings are compatible, their corresponding distribution must be compatible as well. If a binary plan requires its left child to be `AllTuples` and the right child to be `ClusteredDistribution`, I don't think it makes sense to require the children to be co-partitioned. Today, the only case we have is a binary node that requires its children to be both `ClusteredDistribution`.

That said, I think `SinglePartitionShuffleSpec` should only be compatible with `SinglePartitionShuffleSpec`, and `createPartitioning` should take numPartitions and keyIndexes (not an arbitrary `Distribution`) and create partitioning w.r.t. its own `Distribution`.

Reply:
>  If a binary plan requires its left child to be AllTuples and the right child to be ClusteredDistribution, I don't think it makes sense to require the children to be co-partitioned.

I think this is the case of `FlatMapCoGroupsInPandasExec` where one side could be `AllTuples` and the other side could be `ClusteredDistribution`, in which case we should still avoid shuffle if the latter only has one partition? do you mean we should handle it separately, like in `EnsureRequirements` but not in `ShuffleSpec`?

> createPartitioning should take numPartitions and keyIndexes (not an arbitrary Distribution) and create partitioning w.r.t. its own Distribution.

This is sort of what's done at the moment. `createPartitioning` only takes `AllTuples` and `ClusteredDistribution` and it's an error on the implementer side if other distribution types are supplied. Subclasses of `ShuffleSpec` are only required to implement `createPartitioning0` which takes a sequence of expressions (e.g., cluster keys) from the side to be re-shuffled. The numPartitions and keyIndexes are maintained by the shuffle spec (with best parallelism) that is used to re-shuffle the other side. 

For the case:
```sql
SELECT * FROM A JOIN B ON A.c1 = B.d1 AND a.c2 = B.d2
```

If `A` is `HashPartitioning(c1, 10)` and `B` is `HashPartitioning(d2, 5)`, the shuffle spec of `A` will maintain key indexes `{ 0 }`. When re-shuffling `B`, it'll pick `c2` from the input expressions `[c2, d2]` using the indexes and create a partitioning `HashPartitioning(c2, 10)`.

Reply:
> I think this is the case of FlatMapCoGroupsInPandasExec

This is not true. The grouping keys from left and side must match. The semantic of "cogroup" is to find matching data from left and right that have the same value of group keys. It seems like a mistake that we don't check if the grouping keys from left and right are matched or not. cc @HyukjinKwon

Reply:
> When re-shuffling B, it'll pick c2 from the input expressions [c2, d2] using the indexes

How does `B` know the index? Do you mean `ShuffleSpec.createPartitioning` should take another `ShuffleSpec` as a parameter? Then we can get the key indexes from the other shuffle spec.

Reply:
Hm why `B` needs to know the index? we can use `A`'s indexes to pick out the corresponding keys from `B`'s distribution, I think? we pass `B`'s clustering keys from the `ClusteredDistribution`.

Reply:
> don't check if the grouping keys from left and right are matched or not
Yeah I think it should be matched.

Reply:
> we can use A's indexes to pick out the corresponding keys from B's distribution, I think? we pass B's clustering keys from the ClusteredDistribution.

Now I get it. So `ShuffleSpec` takes a subset of its clustering expressions to create a new partitioning. The subset of its clustering expressions is picked according to the key indexes of the other side.

How about the first point? I think the distributions must match if we need the co-partition property. With this restriction, I think the API design can be simplified.

Reply:
I see, do you mean the distributions for both sides of a join could either be all `AllTuples` or all `ClusteredDistribution`?

Yea it may help to simplify the API, and maybe we can handle the `AllTuples` case separately since it's trivial. Let me think about it.

Reply:
Yea, the distribution of children should be of the same type and are compatible (all `AllTuples`, or all `ClusteredDistribution` with the same number of keys, etc.)

Reply:
Got it. Updated.